NitroxClient/MonoBehaviours/MultiplayerMovementController.cs

```diff
@@ -0,0 +1,174 @@
+using NitroxModel.DataStructures;
+using NitroxModel.Packets;
+using NitroxModel_Subnautica.DataStructures;
+using UnityEngine;
+
+namespace NitroxClient.MonoBehaviours;
+
+public class MultiplayerMovementController : MonoBehaviour
+{
+    private static readonly Dictionary<NitroxId, MultiplayerMovementController> movementControllersById = new();
+
+    public float TimeScalar { get; set; } = 1f;
+    public Vector3 TargetPosition { get; set; }
+    public Quaternion TargetRotation { get; set; }
+    public bool Receiving { get; private set; }
+    public bool Broadcasting { get; private set; }
+    public Vector3 Velocity { get; private set; }
+
+    public event Action BeforeUpdate = () => { };
+    public event Action BeforeFixedUpdate = () => { };
+    public event Action AfterUpdate = () => { };
+    public event Action AfterFixedUpdate = () => { };
+
+    private float timeSinceLastBroadcast;
+    private Rigidbody rigidbody;
+    private IPacketSender packetSender;
+    private SimulationOwnership simulationOwnership;
+    private NitroxEntity entity;
+
+    public void SetBroadcasting(bool broadcasting)
+    {
+        Broadcasting = broadcasting;
+        if (Receiving && broadcasting)
+        {
+            Receiving = !broadcasting;
+        }
+    }
+
+    public void SetReceiving(bool receiving)
+    {
+        Receiving = receiving;
+        if (Broadcasting && receiving)
+        {
+            Broadcasting = !receiving;
+        }
+    }
+
+    public static bool TryGetMovementControllerFrom(NitroxId id, out MultiplayerMovementController mc)
+    {
+        if (id == null)
+        {
+            mc = null;
+            return false;
+        }
+
+        if (movementControllersById.TryGetValue(id, out mc))
+        {
+            return mc;
+        }
+
+        if (!NitroxEntity.TryGetObjectFrom(id, out GameObject gameObject))
+        {
+            return false;
+        }
+
+        if (gameObject.TryGetComponent(out mc) && mc)
+        {
+            movementControllersById.Add(id, mc);
+            return true;
+        }
+
+        return false;
+    }
+
+    private static void StartedSimulatingEntity(NitroxId id)
+    {
+        if (TryGetMovementControllerFrom(id, out MultiplayerMovementController mc))
+        {
+            mc.SetBroadcasting(true);
+        }
+    }
+
+    private static void StoppedSimulatingEntity(NitroxId id)
+    {
+        if (TryGetMovementControllerFrom(id, out MultiplayerMovementController mc))
+        {
+            mc.SetReceiving(true);
+        }
+    }
+
+    private void Awake()
+    {
+        packetSender = NitroxServiceLocator.LocateService<IPacketSender>();
+        simulationOwnership = NitroxServiceLocator.LocateService<SimulationOwnership>();
+    }
+
+    private void Start()
+    {
+        rigidbody = GetComponent<Rigidbody>();
+        if (!TryGetComponent(out NitroxEntity nitroxEntity))
+        {
+            NitroxEntity.GenerateNewId(gameObject);
```

Comment:
Should we really generate a new id here? To me it seems more like a log and destroy spot